File: src/main/java/discretemath/structure/sequence/RecurrentSequence.java
package discretemath.structure.sequence;

import discretemath.structure.set.Set;

public abstract class RecurrentSequence<T> extends AbstractSequence<T> {

	public RecurrentSequence(Set<T> codomain) {
		super(codomain);
	}

	@Override
	protected final T getTerm(int n) {
		return getNextRecurrentValue(n);
	}

	protected abstract T getNextRecurrentValue(int n);
}

File: src/main/scala/discretemath/structure/sequence/RecurrentSequence.scala
package discretemath.structure.sequence

import discretemath.structure.set.main._

import scala.math.BigInt

abstract class RecurrentSequence[T](override val codomain: Set[T]) extends AbstractSequence[T](codomain) {

  final override def applyWithoutCheckInput(input: BigInt): T = getNextRecurrentValue(input)

  protected def getNextRecurrentValue(n: BigInt): T
}